File: csharp/robot-name/RobotName.cs
using System;
using System.Collections.Generic;

public class Robot
{
    const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

    private static readonly HashSet<string> generatedNames = new HashSet<string>();

    public Robot()
    {
        Name = GenerateName();
    }

    public string Name { get; private set; }

    public void Reset()
    {
        Name = GenerateName();
    }

    private string GenerateName()
    {
        var name = GenerateRandomName();
        while (generatedNames.Contains(name))
        {
            name = GenerateRandomName();
        }
        
        generatedNames.Add(name);
        return name;

        string GenerateRandomName()
        {
            var random = new Random();
            var letter1 = chars[random.Next(0, chars.Length)];
            var letter2 = chars[random.Next(0, chars.Length)];
            var number = random.Next(1, 1000);
            return $"{letter1}{letter2}{number.ToString("D3")}";
        }
    }
}

File: go/robot-name/robot_name.go
// Package robotname implements a simple library that implements
// method to validate number using luhn formula
package robotname

import (
	"errors"
	"fmt"
	"math/rand"
	"time"
)

var maximumUniqueNames = 26 * 26 * 10 * 10 * 10
var usedNames = make(map[string]bool)

func init() {
	rand.Seed(time.Now().UnixNano())
}

// Robot struct to keep robot name
type Robot struct {
	name string
}

// Name generates random name if not set
func (r *Robot) Name() (string, error) {
	if len(r.name) > 0 {
		return r.name, nil
	}

	if len(usedNames) == maximumUniqueNames {
		return "", errors.New("no more unique names available")
	}

	name := generateRandomName()
	for usedNames[name] {
		name = generateRandomName()
	}

	usedNames[name] = true
	r.name = name
	return r.name, nil
}

// Reset resets Robot name
func (r *Robot) Reset() {
	r.name = ""
}

func generateRandomName() string {
	c1 := rune('A' + rand.Intn(26))
	c2 := rune('A' + rand.Intn(26))
	number := rand.Intn(1000)
	return fmt.Sprintf("%c%c%03d", c1, c2, number)
}
